activerecord/lib/active_record/database_configurations.rb

```diff
@@ -119,12 +119,11 @@ def configs_for(env_name: nil, name: nil, config_key: nil, include_hidden: false
     # If the application has multiple databases +find_db_config+ will return
     # the first DatabaseConfig for the environment.
     def find_db_config(env)
-      configurations
-        .sort_by.with_index { |db_config, i| db_config.for_current_env? ? [0, i] : [1, i] }
-        .find do |db_config|
-          db_config.env_name == env.to_s ||
-            (db_config.for_current_env? && db_config.name == env.to_s)
-        end
+      current_env_configs, other_configs = configurations.partition(&:for_current_env?)
+      [*current_env_configs, *other_configs].find do |db_config|
+        db_config.env_name == env.to_s ||
+          (db_config.for_current_env? && db_config.name == env.to_s)
+      end
```

Comment:
What do you think about eliminating the extra `for_current_env?` calls too?

```suggestion
      env = env.to_s

      configurations.find do |db_config|
        db_config.for_current_env? && (db_config.env_name == env || db_config.name == env)
      end || configurations.find do |db_config|
        db_config.env_name == env
      end
```

That gives an additional boost:

```
=========================== Hundreds of connections ============================

Warming up --------------------------------------
      find_db_config   248.000  i/100ms
 fast_find_db_config   860.000  i/100ms
fast_find_db_config2     2.263k i/100ms
Calculating -------------------------------------
      find_db_config      2.544k (± 2.4%) i/s -     12.896k in   5.071403s
 fast_find_db_config      8.718k (± 5.9%) i/s -     43.860k in   5.051641s
fast_find_db_config2     22.269k (± 2.2%) i/s -    113.150k in   5.083765s

Comparison:
fast_find_db_config2:    22268.7 i/s
 fast_find_db_config:     8717.7 i/s - 2.55x  slower
      find_db_config:     2544.3 i/s - 8.75x  slower
```

Reply:
Nice! Updated.